### ISSUE TITLE
Add ability to specify trustedCA(or use existing from Rancher installation) in Notifier webhook options

### ISSUE THREAD

**Issue by panoslith** [rancher/rancher#24272]
**What kind of request is this (question/bug/enhancement/feature request):**
Feature request

**Steps to reproduce (least amount of steps as possible):**
-Rancher HA, v2.3.2, installed with custom CA certificates.

-When configuring Notifier webhook there is no option to specify trustedCA, hence if the endpoint is SSL terminated, the alert manager will complain with x509: certificate signed by unknown authority.

Currently we don't specify that:
https://github.com/rancher/system-charts/blob/dev/charts/rancher-monitoring/v0.0.5/charts/alertmanager/templates/alertmanager.yaml

https://prometheus.io/docs/alerting/configuration/#http_config

gzrancher/rancher#7311

gzrancher/rancher#8271

gzrancher/rancher#10745

gzrancher/rancher#9991

gzrancher/rancher#8514

gzrancher/rancher#9991

**Comment by briankintz** [comment]
Same as #24263, similar to #17885

**Comment by janeczku** [comment]
The option to specify a trusted CA/certificate should also be supported in the Slack notifier so that users can send alerts via Mattermost or Rocket Chat webhooks.

**Comment by maggieliu** [comment]
Added to the backlog for release consideration, but need to followup with @cloudnautique for detailed requirements

**Comment by aiyengar2** [comment]
This should be supported by Alerting V2, which will allow users to directly modify the AlertManager Config and supply their own http_config.

Issue: https://github.com/rancher/rancher/issues/28348

**Comment by janeczku** [comment]
### Expectation

With current state of Monitoring v2 the following steps are expected to allow a user to configure a custom CA for alert receivers:

1. Enable the monitoring in the dashboard and set the option `Chart Options -> Alerting -> Additional Secrets` to a secret name like "webhook-ca". Note that this secret cannot be precreated before installation since the secret [must exist](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#alertmanagerspec) in the `cattle-monitoring-system` ns which is only created during installation of the app itself.
2. After the app is installed, create the secret `webhook-ca` in the `cattle-monitoring-system` namespace containing a key of "tls.crt" with a value containing the multiline PEM string of the CA certifificate.
3. In order for this CA to be used for any notifiers, the user must then modify the `alertmanager-rancher-monitoring-alertmanager` secret in the `cattle-monitoring-system` namespace and reference the CA file either in the global or per receiver scope. For Example:

```
global:
  resolve_timeout: 5m
  http_config:
    tls_config:
	  ca_file: /etc/alertmanager/secrets/webhook-ca/tls.crt
(...)
```
###  Observations (2.5-alpha4)

**I. Above steps did not work**
Specifically it was not possible to successfully mount the custom CA secret into the alertmanager pod by specifying the name of the secret containing the CA certificate in `Chart Options -> Alerting -> Additional Secrets`. The resulting `alertmanager` CRD object always would be created with the `secrets` option set to `null`.

**II. Above steps seem rather convoluted / not user friendly**

There should be a bettrer user experience making it easy to specify a custom CA. Proposal:

1. When enabling the monitoring, a user should be able to provide a custom CA certificate (as PEM multiline string) as a Chart options.
2. When a such custom CA is specified  the default HTTP client of the alertmanager cluster will automatically be set to use this CA. Meaning the global http_config in `alertmanager-rancher-monitoring-alertmanager` will be configured as above in step 3.

**Comment by aiyengar2** [comment]
#### re: Above steps did not work

On the latest version of the chart, I tested out adding secrets and it worked as expected.

However, I took a different set of steps:
1. I created the `cattle-monitoring-system` namespace
2. I created the secrets `arvind-test` (containing a key that correlates to a YAML file with random data) and `arvind-test-empty` (containing no data)
3. I tried deploying the chart after referencing the secrets within `Chart Options -> Alerting -> Additional Secrets`. 

Once I did this, I was able to see the secrets show up in the default Alertmanager CR that came up and saw the following when I exec-ed into the Alertmanager pod:
```
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ kubectl exec -it alertmanager-rancher-monitoring-alertmanager-0 -c alertmanager -n cattle-monitoring-system ls /etc/alertmanager/secrets                                           
arvind-test             arvind-test-empty.yaml
```

It is possible that the secrets option was empty because, on chart deploy, UI wasn't able to discover the secrets provided within the cluster so they were ignored. In general, I don't think this is a bug; we should expect users to either manually create the namespace and secrets before deploying the chart or deploy the chart and then update it with additional secrets on a chart upgrade.

#### re: Above steps seem rather convoluted / not user friendly

I agree. This primarily is because adding secrets that are referenced by the Alertmanager Config require three steps in succession:

(step 0: create the namespace if it doesn't exist)
1. Add the secret to the `cattle-monitoring-system` namespace
2. Deploy / redeploy the chart with additional secrets
3. Modifying the actual Alertmanager Config YAML on the Secret to point to the additional secret provided

In terms of the UX you proposed, I'm not sure about whether we should provide the ability to supply the custom CA certificate as part of the Chart Deployment options; the issue with that approach is that we only create the Alertmanager Secret on an install but don't touch it on an upgrade / delete as a result of https://github.com/rancher/charts/pull/637.  As a result, a user who creates the Secret on chart deploy would still need to go to the Alertmanager Secret to modify it after the first time the chart is ever deployed; if this was exposed as a field on the UI, we'd also have to add special logic to disable those fields specifically on a Chart upgrade.

Rather than that, I think that users should take the following steps before deploying the chart:

(step 0: create the namespace if it doesn't exist)
1. Add their TLS config secret to the `cattle-monitoring-system` namespace
2. Add their default Alertmanager Config Secret that references that TLS config secret to the `cattle-monitoring-system` namespace (they can just use a boilerplate and then later use the UIs exposed in https://github.com/rancher/dashboard/issues/1238 and https://github.com/rancher/dashboard/issues/1239 to add receivers to the secret).
3. Deploy the chart by referencing those secrets in `Choose an existing config secret` and `Additional Secrets`.

**Comment by janeczku** [comment]
@aiyengar2 

### UX for custom CA cert

> On the latest version of the chart, I tested out adding secrets and it worked as expected. However, I took a different set of steps...

Correct, the approach of creating the namespace and secret in advance works but i don't think it's intuitive / user friendly (see above)

> I'm not sure about whether we should provide the ability to supply the custom CA certificate as part of the Chart Deployment options; the issue with that approach is that we only create the Alertmanager Secret on an install but don't touch it on an upgrade / delete

I think you might be conflating the Alertmanager Config secret with the secret used to store the custom CA certificate: The custom CA certificate would be stored in it's own secret (e.g. `webhook-ca`) which would be lifecycle managed normally through Helm and which user could update/modify by upgrading the chart deployment.

The Alertmanager Config would only contain a reference to the path of the mounted secret like so (pre-install hooks still support templating):

```
global:
  resolve_timeout: 5m
  http_config:
    tls_config:
	  ca_file: /etc/alertmanager/secrets/webhook-ca/tls.crt
(...)
```

Does that make sense?

### UI issue: Additional Secret options

Slightly related, i found an issue with the `Additional Secrets` form field in the UI:
This option must be set to a secret that exists in the `cattle-monitoring-system` namespace as it maps to the `secrets` field of the Alertmanager CRD:

> Secrets [field] is a list of Secrets in the *same* namespace as the Alertmanager object
ref: https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#alertmanagerspec

However the drop-down list is populated with secrets from the entire cluster. It also displays only the resource name without the namespace making it prone to misconfiguration when there are secrets of the same name in multiple namespaces.

The `AlertManager Secret` dropdown list on the contrary is working correctly: It only lists secrets from the `cattle-monitoring-system` namespace.

<img width="1230" alt="Screenshot 2020-09-21 at 22 17 20" src="https://user-images.githubusercontent.com/3813921/93821345-ba3e6c80-fc5e-11ea-983d-565e6ca67d15.png">

**Comment by aiyengar2** [comment]
Thanks for calling that out @janeczku ! That’s a known issue tracked in https://github.com/rancher/dashboard/issues/1230 (under `Additional Secrets have limited namespace`).

**Comment by aiyengar2** [comment]
Regarding the trustedCA UX, the best way I can think of supporting it at the moment would be to do it in two steps:
1) Expose the ability for a user to specify global values to be set in the default Alertmanager Config on chart deployment (this is just editing the YAML within a field in the `values.yaml`; the custom CA certificate can be one of the fields exposed). This field should be editable on a chart install and disabled on any future chart upgrades with a message that links to the Alertmanager Secret for the user to directly modify the secret. @westlywright would this be something we could do?
2) Expose a UI to be able to update the global values of the Alertmanager Config Secret similar to the tickets we have open for Routes and Receivers. This would be a 3rd UI page for Alerting and can be prioritized later.

cc: @deniseschannon @maggieliu

**Comment by janeczku** [comment]
Once the AlertmanagerConfig CRD feature is merged [upstream]( https://github.com/prometheus-operator/prometheus-operator/pull/3451 ) it will be much easier to decouple global alertmanager settings (like an enterprise CA certificate) which should be lifecylce managed by Helm from alerting rules and receivers which a user would maintain out of band of the Helm chart.

The chart would then only manage the `global` part of the Alertmanager Config, including a global CA certificate.
```
global:
  resolve_timeout: 5m
  http_config:
    tls_config:
	  ca_file: /etc/alertmanager/secrets/additional-ca/tls.crt
```
And all routes and receivers would be maintained by user as `AlertmanagerConfig` CRDs.

----

> This field should be editable on a chart install and disabled on any future chart upgrades with a message that links to the Alertmanager Secret for the user to directly modify the secret

I think it's an anti-pattern to allow customization of resources created by Helm only during first install. Resources should either be lifecycle managed by Helm or not at all, in which case it should be obvious to the user that he is responsible to customize and maintain the AlertManager configuration resource out of band of the Helm application.

If we embrace the AlertmanagerConfig CRD in the future, then we could already now add support to maintain the secret resource of a custom CA via the Helm chart. Later we would also manage the `global` part of the AlertManager Config and the reference to that CA secret. Until then, we can instruct the user that this certificate will be available in the AlertManager pod at the particular path `/etc/alertmanager/ca.crt` and how he can modify the AlertManager Config to configure the TLS settings.

**Comment by aiyengar2** [comment]
> Once the AlertmanagerConfig CRD feature is merged upstream it will be much easier to decouple global alertmanager settings (like an enterprise CA certificate) which should be lifecycle managed by Helm from alerting rules and receivers which a user would maintain out of band of the Helm chart.

This should definitely make it easier for users to configure Alertmanager; however, I'm not sure how long the release that contains that PR would take to be introduced into [`kube-prometheus-stack`](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack), which is the upstream chart Monitoring V2 is based on. For reference, the latest release of Prometheus Operator is in `v0.42.1` (released this September), whereas the latest upstream in `kube-prometheus-stack` only support Prometheus Operator version `v0.38.1` (released in April). There are also other CRD changes between those releases that have been introduced (i.e. `Probes`, switching the `apiVersion`, etc.) so I'm not sure if there are any blockers to this change.

> I think it's an anti-pattern to allow customization of resources created by Helm only during first install. Resources should either be fully lifecycle managed by Helm or not at all. In the later case it should be obvious to the user that he is responsible to customize and maintain the AlertManager Config secret directly.

On this same note, I'm concerned about whether the user should be able to configure a custom CA certificate as part of the chart deployment since that means any modifications to that CA certificate would / should be managed as part of Helm as well. Is certificate rotation uncommon for users who use custom CA certificates?

> TLDR; my suggestion for the initial GA would be:
> 1. Add an option to the Helm chart where user can configure a custom CA certificate in PEM format
> 1. In the description of that field make clear that this CA cert will be mounted in the AM pod at a specific path, e.g. /etc/alertmanager/secrets/ca.crt and that the user still needs to configure a http_config in the Alertmanager Config to put it into effect (hyperlink to the upstream configuration).

I'm not too familiar with the use cases for trustedCAs, so just wanted to confirm: this assumes that most users generally have a single custom CA certificate (rather than having unique certificates per receiver) that is either set as part of the global `http_config` or commonly referenced across the `http_config` of multiple receivers?

**Comment by bashofmann** [comment]
> This should definitely make it easier for users to configure Alertmanager; however, I'm not sure how long the release that contains that PR would take to be introduced into kube-prometheus-stack, which is the upstream chart Monitoring V2 is based on. For reference, the latest release of Prometheus Operator is in v0.42.1 (released this September), whereas the latest upstream in kube-prometheus-stack only support Prometheus Operator version v0.38.1 (released in April). There are also other CRD changes between those releases that have been introduced (i.e. Probes, switching the apiVersion, etc.) so I'm not sure if there are any blockers to this change.

In the past, the PRs that I did for this chart were merged quite fast. I expect this to also be the case now that it has moved away from the helm/stable repo. I opened a PR yesterday to update to 0.42.1 (https://github.com/prometheus-community/helm-charts/pull/128). Let's see how it goes :-).

**Comment by maggieliu** [comment]
Closing this since there is a way to set the trustCA in 2.5's new monitoring/alerting feature as mentioned [here](https://github.com/rancher/rancher/issues/24272#issuecomment-696299391). We have opened a new ticket to improve the UX in the future [here](https://github.com/rancher/dashboard/issues/1532).
